File: flutter/app/ios/Runner/BenchMarker.swift
import Flutter
import Foundation
import os

/// Writes the benchmark markers described in AGENTS.md to the platform log.
///
/// Dart takes the timestamps and hands them over on `bench/marker`; only
/// `processStart` is taken here, because Dart cannot see it.
enum BenchMarker {
  static let channelName = "bench/marker"

  private static let logger = Logger(subsystem: "bench", category: "marker")

  static func mark(_ name: String, epochMs: Int64) {
    // Interpolated values are redacted as <private> in release builds unless
    // they are explicitly public.
    logger.notice("BENCH|\(name, privacy: .public)|\(epochMs, privacy: .public)")
    // On a physical iPhone the app log carries only the process's own output,
    // not unified logging, so the line goes to stderr as well.
    fputs("BENCH|\(name)|\(epochMs)\n", stderr)
  }

  /// When the kernel started this process, which is where a cold start begins.
  static func processStartEpochMs() -> Int64? {
    var info = kinfo_proc()
    var size = MemoryLayout<kinfo_proc>.stride
    var mib: [Int32] = [CTL_KERN, KERN_PROC, KERN_PROC_PID, getpid()]
    guard sysctl(&mib, u_int(mib.count), &info, &size, nil, 0) == 0 else {
      return nil
    }
    let start = info.kp_proc.p_un.__p_starttime
    let seconds = TimeInterval(start.tv_sec) + TimeInterval(start.tv_usec) / 1_000_000
    return Int64((seconds * 1000).rounded())
  }

  /// Answers the Dart side's calls. `markLaunch` also emits `processStart`.
  static func register(with messenger: FlutterBinaryMessenger) {
    let channel = FlutterMethodChannel(name: channelName, binaryMessenger: messenger)
    channel.setMethodCallHandler { call, result in
      guard
        let arguments = call.arguments as? [String: Any],
        let name = arguments["name"] as? String,
        let epochMs = arguments["epochMs"] as? NSNumber
      else {
        result(FlutterMethodNotImplemented)
        return
      }
      if call.method == "markLaunch", let start = processStartEpochMs() {
        mark("processStart", epochMs: start)
      }
      mark(name, epochMs: epochMs.int64Value)
      result(nil)
    }
  }
}

File: qr/flutter/app/android/app/src/main/kotlin/com/example/benchmark/qr/qrflutterapp/BenchMarker.kt
package com.example.benchmark.qr.qrflutterapp

import android.os.Process
import android.os.SystemClock
import android.util.Log
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel

/**
 * Writes the benchmark markers described in AGENTS.md to logcat.
 *
 * Dart takes the timestamps and hands them over on `bench/marker`; only
 * `processStart` is taken here, because Dart cannot see it.
 */
object BenchMarker {
  const val CHANNEL_NAME = "bench/marker"

  private const val TAG = "Bench"

  fun mark(name: String, epochMs: Long) {
    Log.i(TAG, "BENCH|$name|$epochMs")
  }

  /** When this process started, converted from uptime to wall-clock time. */
  fun processStartEpochMs(): Long =
    System.currentTimeMillis() - (SystemClock.uptimeMillis() - Process.getStartUptimeMillis())

  /** A figure rather than a time; see AGENTS.md. */
  fun markValue(name: String, value: String) {
    Log.i(TAG, "BENCHVAL|$name|$value")
  }

  /** Answers the Dart side's calls. `markLaunch` also emits `processStart`. */
  fun register(messenger: BinaryMessenger) {
    MethodChannel(messenger, CHANNEL_NAME).setMethodCallHandler { call, result ->
      val name = call.argument<String>("name")
      if (name == null) {
        result.notImplemented()
        return@setMethodCallHandler
      }
      if (call.method == "markValue") {
        markValue(name, call.argument<String>("value") ?: "")
        result.success(null)
        return@setMethodCallHandler
      }
      val epochMs = call.argument<Number>("epochMs")
      if (epochMs == null) {
        result.notImplemented()
        return@setMethodCallHandler
      }
      if (call.method == "markLaunch") {
        mark("processStart", processStartEpochMs())
      }
      mark(name, epochMs.toLong())
      result.success(null)
    }
  }
}
